### PR TITLE
Optimize tokens

### PR DESCRIPTION
**Pull request by six7**

### CODE REVIEW

File: src/app/components/TokenListing.tsx [hunk 1]
```diff
@@ -106,30 +115,16 @@ const TokenListing: React.FC<Props> = ({
...
-              type="button"
-              onClick={() => {
-                showNewForm({});
-              }}
-            >
-              <Icon name="add" />
-            </button>
-          </Tooltip>
+          <IconButton
+            data-cy="button-add-new-token"
```
Comment (six7) [inline src/app/components/TokenListing.tsx:122]:
```suggestion
            dataCy="button-add-new-token"
```

File: src/app/components/TokenTree.tsx [hunk 2]
```diff
@@ -59,17 +59,16 @@ const TokenTree: React.FC<Props> = ({
...
-                <Tooltip label="Add a new token" side="left">
-                  <button
+                <div className="opacity-0 group-hover:opacity-100 focus:opacity-100">
+                  <IconButton
+                    icon={<AddIcon />}
+                    tooltip="Add a new token"
+                    tooltipSide="left"
+                    onClick={handleShowNewForm}
                     disabled={editProhibited}
                     data-cy="button-add-new-token-in-group"
```
Comment (six7) [inline src/app/components/TokenTree.tsx:69]:
```suggestion
                    dataCy="button-add-new-token-in-group"
```

File: src/app/components/ImportedTokensDialog.tsx [hunk 3]
```diff
@@ -75,7 +69,15 @@ export default function ImportedTokensDialog() {
   const [newTokens, setNewTokens] = React.useState(importedTokens.newTokens);
   const [updatedTokens, setUpdatedTokens] = React.useState(importedTokens.updatedTokens);
 
-  const handleCreateNewClick = () => {
+  const handleSetUpdatedTokens = React.useCallback((tokens: SingleToken[]) => {
```
Comment (LiamMartens) [inline src/app/components/ImportedTokensDialog.tsx:72]: @six7 good to see useCallbacks! :smile:

File: src/app/components/ImportedTokensDialog.tsx [hunk 4]
```diff
@@ -192,8 +194,8 @@ export default function ImportedTokensDialog() {
                 value={token.value}
                 description={token.description}
                 updateAction="Create"
-                removeToken={() => setNewTokens(newTokens.filter((t) => t.name !== token.name))}
-                updateToken={() => handleCreateSingleClick(token)}
+                removeToken={handleSetNewTokens(newTokens.filter((t) => t.name !== token.name))}
+                updateToken={handleCreateSingleClick(token)}
```
Comment (LiamMartens) [inline src/app/components/ImportedTokensDialog.tsx:198]: Did you mean to write these like this? Now the handlers will be executed on render, which presumably is not the intention. The handlers itself should just use the `token` and `newTokens` variables and then you can just pass the functions into the props as is `removeToken={handleSetewTokens}`.

File: src/app/components/ImportedTokensDialog.tsx [hunk 5]
```diff
@@ -224,8 +226,8 @@ export default function ImportedTokensDialog() {
                 description={token.description}
                 oldDescription={token.oldDescription}
                 updateAction="Update"
-                removeToken={() => setUpdatedTokens(updatedTokens.filter((t) => t.name !== token.name))}
-                updateToken={() => handleUpdateSingleClick(token)}
+                removeToken={handleSetUpdatedTokens(updatedTokens.filter((t) => t.name !== token.name))}
```
Comment (LiamMartens) [inline src/app/components/ImportedTokensDialog.tsx:229]: Same thing as above

**Review by LiamMartens — CHANGES_REQUESTED** [review]

File: src/app/components/ImportedTokensDialog.tsx (hunk 4 above)
Comment (six7) [reply src/app/components/ImportedTokensDialog.tsx:221]: Thanks, hope I got that right with this one. The handler itself now uses the token, however I still have to pass the current token in, like this:

```
  const handleIgnoreExistingToken = React.useCallback((token) => () => {
    setUpdatedTokens((updatedTokens.filter((t) => t.name !== token.name)));
  }, [setUpdatedTokens, updatedTokens]);

  ...
  removeToken={handleIgnoreExistingToken(token)}
```

Comment (LiamMartens) [reply src/app/components/ImportedTokensDialog.tsx:221]: Then this actually needs to be separated into it's own component taking the `token` as a prop. Then you can create handlers with those tokens specifically.

**Review by LiamMartens — CHANGES_REQUESTED** [review]

File: src/app/components/ImportedTokensDialog.tsx (hunk 4 above)
Comment (six7) [reply src/app/components/ImportedTokensDialog.tsx]: Got it, fixed in newest commit. 🙏

File: src/types/tokens/ImportToken.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import { SingleToken } from './SingleToken';
+
+export type ImportToken = SingleToken & {
```
Comment (LiamMartens) [inline src/types/tokens/ImportToken.ts:3]: Could you just change this with
```
export type ImportToken<Named = true> = SingleToken<Named, {
  oldValue?: string;
  oldDescription?: string;
}>
```

Keeps the typing more strict

**Review by LiamMartens — APPROVED** [review]
@six7 just one small comment - other than that good to go

**Pull request merged**